Tests/MLXTests/MLXArray+IndexingTests.swift

```diff
@@ -348,4 +368,276 @@ class MLXArrayIndexingTests: XCTestCase {
+        check(
+            a[.stride(by: -1), ..<2, 2..., .ellipsis, .newAxis, .stride(by: 2)],
+            [2, 2, 2, 5, 1, 2], 13160)
+    }
+
+    public func testFullIndexReadArray() {
+        // these have an MLXArray as a source of indexes and go through the gather path
+
+        // a = mx.arange(540).reshape(3, 3, 4, 5, 3)
+        let a = MLXArray(0 ..< 540, [3, 3, 4, 5, 3])
+
+        // i = mx.array([2, 1])
+        let i = MLXArray([2, 1])
+
+        // a[0, i]
+        check(a[0, i], [2, 4, 5, 3], 14340)
+
+        // a[..., i, 0]
+        check(a[.ellipsis, i, 0], [3, 3, 4, 2], 19224)
+
+        // a[i, 0, ...]
+        check(a[i, 0, .ellipsis], [2, 4, 5, 3], 35940)
+
+        // gatherFirst path
+        // a[i, ..., i]
+        check(a[i, .ellipsis, i], [2, 3, 4, 5], 43200)
+
+        // a[i, ..., ::2, :]
+        check(a[i, .ellipsis, .stride(by: 2), 0...], [2, 3, 4, 3, 3], 77652)
+
+        // gatherFirst path
+        // a[..., i, None, ::2, -1]
+        check(a[.ellipsis, i, .newAxis, stride(by: 2), -1], [2, 3, 3, 1, 3], 14607)
+
+        // a[:, 2:, i]
+        check(a[0..., 2..., i], [3, 1, 2, 5, 3], 29655)
+
+        // a[::-1, :2, i, 2:, ..., None, ::2]
+        check(
+            a[.stride(by: -1), ..<2, i, 2..., .ellipsis, .newAxis, .stride(by: 2)],
+            [3, 2, 2, 3, 1, 2], 17460)
+    }
+
+    // MARK: - Full Indexing - Write
+
+    public func testFullIndexWriteSingle() {
+        // single operators go through an optimized path
+
+        func check(
+            _ indexes: [MLXArrayIndex], _ sum: Int, file: StaticString = #filePath,
+            line: UInt = #line
+        ) {
+            // a = mx.arange(60).reshape(3, 4, 5)
+            let a = MLXArray(0 ..< 60, [3, 4, 5])
+
+            a[operations: indexes.map { $0.mlxArrayIndexOperation }] = MLXArray(1)
+            XCTAssertEqual(a.sum().item(Int.self), sum, file: file, line: line)
+        }
+
+        // a[...]
+        // not valid
+
+        // a[None]
+        check([.newAxis], 60)
+
+        // a[0]
+        check([0], 1600)
+
+        // a[1:3]
+        check([1 ..< 3], 230)
+
+        // i = mx.array([2, 1])
+        let i = MLXArray([2, 1])
+
+        // a[i]
+        check([i], 230)
+    }
+
+    public func testFullIndexWriteNoArray() {
+        func check(
+            _ indexes: [MLXArrayIndex], _ sum: Int, file: StaticString = #filePath,
+            line: UInt = #line
+        ) {
+            // a = mx.arange(360).reshape(2, 3, 4, 5, 3)
+            let a = MLXArray(0 ..< 360, [2, 3, 4, 5, 3])
+
+            a[operations: indexes.map { $0.mlxArrayIndexOperation }] = MLXArray(1)
+            XCTAssertEqual(a.sum().item(Int.self), sum, file: file, line: line)
+        }
+
+        // a[..., 0] = 1
+        check([.ellipsis, 0], 43320)
+
+        // a[0, ...] = 1
+        check([0, .ellipsis], 48690)
+
+        // a[0, ..., 0] = 1
+        check([0, .ellipsis, 0], 59370)
+
+        // a[..., ::2, :] = 1
```

Comment:
See also https://github.com/ml-explore/mlx/issues/937 -- the swift code has the same bug, but I applied my suggested workaround and tested it vs numpy.